AK/PrintfImplementation.h

```diff
@@ -283,166 +285,227 @@ ALWAYS_INLINE int print_signed_number(PutChFunc putch, char*& bufptr, int number
     return print_number(putch, bufptr, number, left_pad, zero_pad, field_width);
 }
 
-template<typename PutChFunc>
-ALWAYS_INLINE int printf_internal(PutChFunc putch, char* buffer, const char*& fmt, va_list ap)
+struct State {
+    bool left_pad = false;
+    bool zero_pad = false;
+    bool dot = false;
+    unsigned field_width = 0;
+    bool has_fraction_length = false;
+    unsigned fraction_length = 6;
+    unsigned long_qualifiers = 0;
+    bool size_qualifier = false;
+    bool alternate_form = 0;
+    bool always_sign = false;
+};
+
+template<typename PutChFunc, typename ArgumentListT, template<typename T, typename U = ArgumentListT> typename NextArgument>
```

Comment:
I still don't see why that `U` is needed.

Reply:
Since `va_list&` cannot be named, it has to be passed as a type parameter, making the template type of `NextArgument` require two params; but the second parameter is just the input argument list type.

in short, this is here to make it possible to pass `va_list` as a reference.

Reply:
What do you mean it cannot be named?

Reply:
Try making a function take a `va_list&`, and pass a `va_list` into it.
you cannot name it in a normal way, array type and all that.

Reply:
Oh, that is simple to workaround, just wrap it in a struct.

Reply:
Like this:

```cpp
struct { va_list va_list; } args;
va_copy(args.va_list, original_args);
```

Reply:
Then _everything_ will have to be wrapped in a struct.
or we'll have to do some SFINAE trickery to selectively wrap `va_list`'s
Also, can you even move a `va_list`? how would wrapping it even work?

```c++
struct Foo {
    va_list list;
};

va_list ap;
Foo x { move(ap) }; // This doesn't seem to make much sense to me
```

Reply:
> Then everything will have to be wrapped in a struct. or we'll have to do some SFINAE trickery to selectively wrap va_list's

No? Just the implementation of `NextArgument` on to of `va_list` needs to care. The rest, and wasn't this the whole point, is oblivious to how exactly you're getting the argument values.

Reply:
> `va_copy(args.va_list, original_args);`

That will either require the ctor of that struct to take a `va_list` as a reference (defeating the point), or the instantiating site to specialise (probably via `if constexpr`) for `va_list`.
Not a fan of either, but if the template args method is really much worse than either of these, I'll swap it out

Reply:
> No? Just the implementation of NextArgument on top of va_list needs to care.

I missed that entirely somehow. _that's_ reasonable.

Reply:
I tried doing this (conditionally wrapping the `va_list`), but it just doesn't want to cooperate.
I don't think losing one template template parameter is worth-
- The wrapping cost (if any)
- The number of new structures added to make this conditional wrapping possible